### PR TITLE
Implement laser envelope model in vacuum

### PR DESCRIPTION
**Pull request by PierrePelletier**
This is a first draft for implementing an envelope model for the laser, which would allow for a lower resolution along the z axis.
In this PR we modified heavily fields.py and its dependencies to allow for the handling of the envelope of the laser. We thus store the A envelope field, as well as the dtA field (derivative in time) and we are able to push them forward in time. For this, as the envelope is inherently complex, we also need to compute the negative azimuthal modes and so we store 2Nm - 1 modes in total. In order to do this, the InterpolationGrid and SpectralGrid objects have each been separated in two variations, Envelope and Field which contains everything the old grids used to hold.

The simulation is for now capable of simulating the propagation of the A envelope field in a vacuum. New methods have also been created to initialize the field created by a Gaussian laser as an envelope A field rather than a standard E field. There is also a test file added allowing to test the propagation of the envelope in a vacuum, currently only for m = 0.

The following has not yet been implemented:
-Any interactions with the plasma
-MPI or GPU use 
-Comoving approach (Note: the envelope model as of now is not valid for high boosts)
-Handling moving windows
-Boundary conditions
-Any further tests

### CODE REVIEW

**Comment by RemiLehe** [comment]
From offline discussions with @PierrePelletier: this PR will be reopened after implementing a mechanism to initialize the laser envelope only when needed. (right now the envelope is used by default)

**Comment by PierrePelletier** [comment]
Reopening this PR since the mechanism to initialize the laser envelope has been implemented.

This mechanism work by setting a flag in the Field object when you call the method adding the laser envelope to the mesh.
On top of that, the code has now been tested for mode m = 1 with a LaguerreGaussian profile

**Comment by RemiLehe** [comment]
Great! Thanks for reopening the PR. I'll start reviewing it tomorrow.

File: fbpic/fields/fields.py [hunk 1]
```diff
@@ -446,7 +532,7 @@ def partial_interp2spect(self, fieldtype) :
         ---------
         fieldtype :
             A string which represents the kind of field to transform
-            (either 'E', 'B', 'J', 'rho_next', 'rho_prev')
+            (either 'E', 'B', 'J', 'rho_next', 'rho_prev’, ‘A’)
```
Comment (RemiLehe) [inline fbpic/fields/fields.py]: Could you replace the back ticks here (`) by regular ticks (')?
I think that this is what is making the Python2 tests fail.

File: fbpic/fields/fields.py [hunk 2]
```diff
@@ -571,3 +673,15 @@ def divide_by_volume( self, fieldtype ) :
         """
         for m in range(self.Nm):
             self.interp[m].divide_by_volume( fieldtype )
+
```
Comment (RemiLehe) [inline fbpic/fields/fields.py]: If we create a list `self.envelope_mode_numbers`, then this method can be suppressed.

File: fbpic/fields/interpolation_grid.py [hunk 3]
```diff
@@ -70,18 +70,6 @@ def __init__(self, Nz, Nr, m, zmin, zmax, rmax, use_cuda=False ) :
         # Note: No Verboncoeur-type correction required
         self.invvol = 1./vol
```
Comment (RemiLehe) [inline fbpic/fields/interpolation_grid.py]: Could you update the docstring of the `InterpolationGrid` to say that this is a base class, which is used for the `FieldInterpolationGrid` and `EnvelopeInterpolationGrid`?

File: fbpic/fields/fields.py [hunk 4]
```diff
@@ -496,12 +590,16 @@ def erase(self, fieldtype ) :
         ---------
         fieldtype : string
             A string which represents the kind of field to be erased
-            (either 'E', 'B', 'J', 'rho')
+            (either 'E', 'B', 'J', 'rho' or 'A')
```
Comment (RemiLehe) [inline fbpic/fields/fields.py]: For now, since we do not need the `erase` method, I would recommend reverting it to its previous version (without A)

File: fbpic/fields/fields.py [hunk 5]
```diff
@@ -549,11 +647,15 @@ def filter_spect( self, fieldtype ) :
         ---------
         fieldtype : string
             A string which represents the kind of field to be filtered
-            (either 'E', 'B', 'J', 'rho_next' or 'rho_prev')
+            (either 'E', 'B', 'J', 'rho_next', 'rho_prev' or 'A')
```
Comment (RemiLehe) [inline fbpic/fields/fields.py]: For now, since we do not need the `filter` method yet, I would recommend to revert it to its previous version (without A)

File: fbpic/fields/interpolation_grid.py [hunk 6]
```diff
@@ -218,3 +229,24 @@ def divide_by_volume( self, fieldtype ) :
...
+class EnvelopeInterpolationGrid(InterpolationGrid):
+    
+    def __init__(self, Nz, Nr, m, zmin, zmax, rmax, use_cuda=False ) :
+        
+        InterpolationGrid.__init__(self, Nz, Nr, m, zmin, zmax, rmax, use_cuda=False)
+        
+        # Allocate the fields arrays
+        self.A = np.zeros( (Nz, Nr), dtype='complex' )
+        self.dtA = np.zeros( (Nz, Nr), dtype='complex' )
+
```
Comment (RemiLehe) [inline fbpic/fields/interpolation_grid.py]: For now, the `erase` method is not needed, and so I would recommend to just remove it for the moment.

File: fbpic/fields/numba_methods.py [hunk 7]
```diff
@@ -139,7 +139,26 @@ def numba_push_eb_standard( Ep, Em, Ez, Bp, Bm, Bz, Jp, Jm, Jz,
                 + j_coef[iz, ir]*( 1.j*kr[iz, ir]*Jp[iz, ir] \
                             + 1.j*kr[iz, ir]*Jm[iz, ir] )
 
+
     return
+    
+@njit_parallel
+def numba_push_envelope_standard(A, dtA, w2_square, invw_tot, S_env, C_env,
+                            sinc_env, A_coef, Nz, Nr):
```
Comment (RemiLehe) [inline fbpic/fields/numba_methods.py]: Could you add a docstring for this function? (on the same model as for the `push_eb_standard`)

File: fbpic/fields/psatd_coefs.py [hunk 8]
```diff
@@ -173,3 +173,37 @@ def __init__( self, kz, kr, m, dt, Nz, Nr, V=None,
...
+        dt : float
+            The timestep of the simulation
+        
+        k0: float
+            Wavenumber of the beam represented by the envelope model
+        """
+        
+        # Calculate all necessary coefficients for propagation of A field
+        w_laser = c * k0
+        self.w2_square = c**2 * (kr**2 + kz**2 + 2 * kz * w_laser/c)
```
Comment (RemiLehe) [inline fbpic/fields/psatd_coefs.py]: Could you just use `k0` instead of `w_laser/c` in the above expression?

File: fbpic/fields/spectral_grid.py [hunk 9]
```diff
@@ -410,3 +432,48 @@ def filter(self, fieldtype) :
...
+            ps : PsatdCoeffs object
+                psatd object corresponding to the same m mode
+            """
+            assert (ps.V is None or ps.V == 0)
+            assert( abs(self.m) == ps.m )
+        
+            numba_push_envelope_standard(self.A, self.dtA, ps.w2_square, ps.invw_tot, ps.S_env,
+                ps.C_env, ps.sinc_env, ps.A_coef, self.Nz, self.Nr)
+
+
```
Comment (RemiLehe) [inline fbpic/fields/spectral_grid.py]: For now, I think that the `filter` function is not needed, and so it could be removed.

File: fbpic/fields/psatd_coefs.py [hunk 10]
```diff
@@ -173,3 +173,37 @@ def __init__( self, kz, kr, m, dt, Nz, Nr, V=None,
...
+        w_tot = np.sqrt( (w_laser + c * kz)**2 + c**2 * kr**2)
```
Comment (RemiLehe) [inline fbpic/fields/psatd_coefs.py]: Instead of `invw_tot` and `S_env` could we actually store their product (since this is what is used in the actual push function)?
e.g.
```
self.S_env_over_w = np.sin(w_tot*dt) / np.where( w_tot == 0, 1, w_tot )
self.S_env_over_w[ w_tot==0 ] = dt
```
This would have the advantage of computing the correct value for the limit `w_tot = 0`.

Of the course the `push_envelope` function would have to be modified.

File: tests/unautomated/test_laser_envelope.py [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+# Parameters
+# ----------
+# (See the documentation of the function propagate_pulse
+# below for their definition)
+show = False # Whether to show the plots, and check them manually
+
+use_cuda = False
+
+# Simulation box
+Nz = 400
```
Comment (RemiLehe) [inline tests/unautomated/test_laser_envelope.py]: Actually, the envelope solver should work also when the laser wavelength is badly resolved ; therefore, I am wondering if using a small `Nz` (e.g. `Nz = 80`) would still work for this test. 
If yes, it would be great to modify this parameter accordingly.

File: tests/unautomated/test_laser_envelope.py [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+zmax = 10.e-6
+Nr = 25
+Lr = 40.e-6
+Nm = 2
+n_order = -1
+# Laser pulse
+w0 = 4.e-6
+ctau = 5.e-6
+k0 = 2*np.pi/0.8e-6
+E0 = 1.
```
Comment (RemiLehe) [inline tests/unautomated/test_laser_envelope.py]: Maybe we should replace E0 by a0 in this test.

File: tests/unautomated/test_laser_envelope.py [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+    fld : Fields object from fbpic
+
+    m : int, optional
+       The index of the mode to be fitted
+    """
+    # Integrate the laser oscillations longitudinally
+    dz = fld.interp[0].dz
+    laser_profile = np.sqrt(dz*(abs( fld.envelope_interp[m].A )**2).sum(axis=0)) 
+    # Renormalize so that this gives the peak of the Gaussian
+    laser_profile *= 2.**(-3./4)/( np.pi**(1./4) * ctau**(1./2) )
```
Comment (RemiLehe) [inline tests/unautomated/test_laser_envelope.py]: Actually, the aim of the 3 above lines was originally to evaluate the amplitude of the laser envelope, in the case where the available data is an oscillating electric field.

However, here, the available data is the laser envelope directly. So we could replace these 3 lines by simply
```
laser_envelope = abs( fld.envelope_interp[m].A )
```

File: tests/unautomated/test_laser_envelope.py [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+                    exchange_period = 1, use_galilean=use_galilean )
+    # Remove the particles
+    sim.ptcl = []
+    # Set the moving window object
+    if v_window !=0:
+        sim.set_moving_window( v=v_window )
+
+    # Initialize the laser fields
+    z0 = (zmax+zmin)/2
+    init_fields( sim, w0, ctau, k0, z0, zf, E0, m )
```
Comment (RemiLehe) [inline tests/unautomated/test_laser_envelope.py]: We could again use `a0` instead of `E0` here.

File: tests/unautomated/test_laser_envelope.py [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+    else:
+        assert np.allclose( w, w_analytic, rtol=rtol )
+        assert np.allclose( A, A_analytic, rtol=5.e-3 )
+        print('The simulation results agree with the theory to %e.' %rtol)
+
+    # Return a dictionary of the results
+    return( { 'A' : A, 'w' : w, 'fld' : sim.fld } )
+
+
+def init_fields( sim, w, ctau, k0, z0, zf, E0, m=1 ) :
```
Comment (RemiLehe) [inline tests/unautomated/test_laser_envelope.py]: We could again use `a0` instead of `E0` here.

File: fbpic/lpa_utils/laser/direct_envelope_injection.py [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+    # Check that no other laser simulation has been already added
+    if sim.fld.use_envelope:
+        raise ValueError("Another laser profile has already been added")
+    sim.fld.activate_envelope_model(laser_profile.k0)
+
+    # Get the local azimuthally-decomposed laser fields A and dtA on each proc
+    laser_A, laser_dtA = get_laser_A_dtA( sim, laser_profile, boost )
+    for m in range(2 * sim.fld.Nm - 1):
+        sim.fld.envelope_interp[m].A[:,:] = laser_A[:,:,m]
+        sim.fld.envelope_interp[m].dtA[:,:] = laser_dtA[:,:,m]
```
[the remark this answers is not in the document]
Comment (PierrePelletier) [reply fbpic/lpa_utils/laser/direct_envelope_injection.py]: Oh right thank you for that I was mistaken on the number of modes that would be calculated. That also explains the troubles I had when testing for m = -1. 
But then why do we use ntheta = 2 * Nm when we only study 2* Nm -1 modes (even if studied simultaneously for the real fields)?

File: tests/unautomated/test_laser_envelope.py (hunk 13 above)
Comment (PierrePelletier) [reply tests/unautomated/test_laser_envelope.py]: I'm pretty sure this method wouldn't work since we actually need to make the z dependency disappear in the array.
I just tried evaluating the envelope at the estimated center of the beam in the z direction then fit the curve ( something like laser_profile = abs(fld.envelope_interp[m].A[izref]) with izref well chosen) which seems to work pretty well

File: fbpic/fields/interpolation_grid.py [hunk 17]
```diff
@@ -99,6 +87,42 @@ def r(self):
...
+    
+    def __init__(self, Nz, Nr, m, zmin, zmax, rmax, use_cuda=False ) :
+        """
+        Initialize a 'FieldInterpolationGrid' object
+
+        See the docstring of the parent class 'InterpolationGrid' 
+        for the meaning of the different parameters.
+        """
+        
+        InterpolationGrid.__init__(self, Nz, Nr, m, zmin, zmax, rmax, use_cuda=False)
```
Comment (RemiLehe) [inline fbpic/fields/interpolation_grid.py]: Could you pass `use_cuda=use_cuda` instead of `use_cuda=False`?

File: fbpic/fields/interpolation_grid.py [hunk 18]
```diff
@@ -218,3 +236,24 @@ def divide_by_volume( self, fieldtype ) :
...
+    """
+    Contains the coordinates and envelope of the spatial grid.
+    
+    Main attributes:
+    - z,r : 1darrays containing the positions of the grid
+    - A, dtA:
+      2darrays containing the envelope amplitude.
+    """
+    
+    def __init__(self, Nz, Nr, m, zmin, zmax, rmax, use_cuda=False ) :
```
[the remark this answers is not in the document]
Comment (RemiLehe) [reply fbpic/fields/interpolation_grid.py]: Could you also pass `use_cuda=use_cuda` here?

File: fbpic/fields/interpolation_grid.py [hunk 19]
```diff
@@ -218,3 +242,30 @@ def divide_by_volume( self, fieldtype ) :
...
+        """
+        Initialize a 'EnvelopeInterpolationGrid' object
+
+        See the docstring of the parent class 'InterpolationGrid' 
+        for the meaning of the different parameters.
+        """
+        
+        InterpolationGrid.__init__(self, Nz, Nr, m, zmin, zmax, rmax, use_cuda=False)
```
Comment (RemiLehe) [inline fbpic/fields/interpolation_grid.py]: Could you also pass `use_cuda=use_cuda` here?

File: fbpic/fields/spectral_grid.py [hunk 20]
```diff
@@ -112,10 +98,62 @@ def __init__(self, kz_modified, kr, m, kz_true, dz, dr,
...
+    def __init__(self, kz_modified, kr, m, kz_true, dz, dr,
+                        current_correction, use_cuda=False ) :
+        """
+        Initialize a 'FieldSpectralGrid' object
+
+        See the docstring of the parent class 'SpectralGrid' 
+        for the meaning of the different parameters.
+        """
+        SpectralGrid.__init__(self, kz_modified, kr, m, kz_true, dz, dr,
+                        use_cuda=False )
```
Comment (RemiLehe) [inline fbpic/fields/spectral_grid.py]: Same comment on use_cuda

File: fbpic/fields/spectral_grid.py [hunk 21]
```diff
@@ -410,3 +449,54 @@ def filter(self, fieldtype) :
...
+                        use_cuda=False ) :
+        """
+        Initialize a 'EnvelopeSpectralGrid' object
+
+        See the docstring of the parent class 'SpectralGrid' 
+        for the meaning of the different parameters.
+        """
+        
+        SpectralGrid.__init__(self, kz_modified, kr, m, kz_true, dz, dr,
+                        use_cuda=False )
```
Comment (RemiLehe) [inline fbpic/fields/spectral_grid.py]: Same comment on use_cuda
